KOGITO-3716 Allow automatic close of staging repo

radtriste opened the pull request:
https://issues.redhat.com/browse/KOGITO-3716

Dependent PRs:
- https://github.com/kiegroup/kogito-pipelines/pull/91
- https://github.com/kiegroup/kogito-runtimes/pull/865
- https://github.com/kiegroup/optaplanner/pull/1010
- https://github.com/kiegroup/kogito-examples/pull/435
- https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/114

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [ ] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

radtriste commented on Jenkinsfile.deploy:
```diff
@@ -270,9 +273,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
+
+        // Deploy to staging repository and close
+        cmd = 'org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:deploy-staged-repository -DkeepStagingRepositoryOnCloseRuleFailure=true -DstagingProgressTimeoutMinutes=10'
+        cmd += "--projects ${parentProjectName ?: directory}"
```
```suggestion
        cmd += " --projects ${parentProjectName ?: directory}"
```

radtriste commented on Jenkinsfile.deploy:
```diff
@@ -270,9 +273,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
+
+        // Deploy to staging repository and close
+        cmd = 'org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:deploy-staged-repository -DkeepStagingRepositoryOnCloseRuleFailure=true -DstagingProgressTimeoutMinutes=10'
+        cmd += " --projects ${parentProjectName ?: directory}"
```
```suggestion
        cmd += " --projects :${parentProjectName ?: directory}"
```

radtriste commented on Jenkinsfile.deploy:
```diff
@@ -270,9 +273,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
+
+        // Promote to build profile
+        cmd = "org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:promote -DbuildPromotionProfileId=${env.NEXUS_BUILD_PROMOTION_PROFILE_ID}"
+        cmd += "--projects ${directory}"
```
```suggestion
        cmd += " --projects :${parentProjectName ?: directory}"
```

radtriste commented:
Almost there but testing is blocked by https://issues.redhat.com/browse/KOGITO-3795 ...

radtriste commented:
Jenkins retest this

radtriste commented:
Jenkins retest this

Ginxo commented on Jenkinsfile.deploy:
```diff
@@ -203,30 +224,12 @@ void prepareForPR(String repo) {
-        addNotIgnoredPoms()
-        sh "git commit -m '${commitMsg}'"
+        def commitMsg = "Update project version to ${getProjectVersion()} for release"
+        def prBody = "Generated by build ${BUILD_TAG}: ${BUILD_URL}.\nPlease do not merge, it will be merged automatically after testing."
```
what if you replace it by
```
def prBody = """
Generated by build ${BUILD_TAG}: ${BUILD_URL}.
Please do not merge, it will be merged automatically after testing.
"""
```

radtriste replied:
This is a good idea :) I guess I will implement it as a nice to have in another PR because I would like that one to be merged asap ;)
Thx for suggestion !

Ginxo reviewed (approved):
LGTM, just check my proposal in case you like it

rsynek commented on Jenkinsfile.deploy:
```diff
@@ -24,13 +32,20 @@ pipeline {
+        // Nexus staging default parameters
+        string(name: 'NEXUS_RELEASE_URL', defaultValue: 'https://repository.jboss.org/nexus', description: 'Nexus URL for release staging')
+        string(name: 'NEXUS_RELEASE_REPOSITORY_ID', defaultValue: 'jboss-releases-repository', description: 'Nexus Release repository ID for staging')
+        string(name: 'NEXUS_STAGING_PROFILE_ID', defaultValue: '2161b7b8da0080', description: 'Nexus staging profile ID for release process ')
+        string(name: 'NEXUS_BUILD_PROMOTION_PROFILE_ID', defaultValue: 'ea49ccd6f174', description: 'Nexus Build Promotion profile ID for release process')
```
Are these two IDs constant between releases or do we have to provide new ones per release?

radtriste replied:
those are constant. but for testing purpose, there might be different

rsynek commented on Jenkinsfile.deploy:
```diff
@@ -142,8 +152,19 @@ pipeline {
-                mavenDeploy('kogito-runtimes')
-                mavenDeploy('kogito-apps')
+                script {
+                    if(!isRelease() || params.MAVEN_DEPLOY_REPOSITORY){
```
Formatting:

```suggestion
                    if(!isRelease() || params.MAVEN_DEPLOY_REPOSITORY) {
```

rsynek commented on Jenkinsfile.deploy:
```diff
@@ -142,8 +152,19 @@ pipeline {
-                mavenDeploy('kogito-runtimes')
-                mavenDeploy('kogito-apps')
+                script {
+                    if(!isRelease() || params.MAVEN_DEPLOY_REPOSITORY){
```
If the `params.MAVEN_DEPLOY_REPOSITORY` is used only for testing, a comment would be nice.

radtriste replied:
done

rsynek commented on Jenkinsfile.promote:
```diff
@@ -284,28 +279,11 @@ void prepareForPR(String repo) {
-    addNotIgnoredPoms()
-    sh "git commit -m '${commitMsg}'"
+    
+    githubscm.commitChanges(commitMsg, { githubscm.findAndStageNotIgnoredFiles('pom.xml') })
```
Nice to see this replacing the `addNotIgnoredPoms()`.

radtriste commented:
@Kevin-Mok I removed the jenkins tests from main pom and instead have its own workflow (via GHActions)
There was a problem with importing MavenCommand from shared libs in pipeline so I changed it to first retrieve the shared libs

radtriste commented:
squashed commits

rsynek commented on .jenkins/tests/src/test/vars/JenkinsfilePromote.groovy:
```diff
@@ -4,8 +4,8 @@ class JenkinsfilePromote extends JenkinsPipelineSpecification {
     def setup() {
-        Jenkinsfile = loadPipelineScriptForTest('Jenkinsfile.promote')
-        Jenkinsfile.getBinding().setVariable('PROPERTIES_FILE_NAME', 'deployment.properties')
+			Jenkinsfile = loadPipelineScriptForTest('Jenkinsfile.promote')
```
Was this formatting change intentional?

radtriste replied:
nope, it is my F**** IDE using real tabs instead of whitespaces ... gonna correct that ...

radtriste commented:
@rsynek all good to be merged !
